bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,230 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+		 * eg: zebra deallocated the labels and notifies it
+		 */
+		blnc->label = MPLS_INVALID_LABEL;
+
+	if (old_label == blnc->label)
+		return 0; /* no change */
+
+	/* update paths */
+	if (blnc->label != MPLS_INVALID_LABEL)
+		bgp_zebra_send_nexthop_label(ZEBRA_MPLS_LABELS_ADD, blnc->label,
+					     ZEBRA_LSP_BGP, &blnc->nexthop);
+
+	LIST_FOREACH (pi, &(blnc->paths), label_nh_thread) {
+		if (!pi->net)
+			continue;
+		table = bgp_dest_table(pi->net);
+		if (!table)
+			continue;
+		vpn_leak_from_vrf_update(blnc->to_bgp, table->bgp, pi);
+	}
+
+	return 0;
+}
+
+/* get label for a given bgp update
+ * return the label value:
+ * - per-vrf value:
+ *      if nexthop of path is not valid or there are no nexthops
+ *      if nexthop of path is not connected
+ *      if path becomes ECMP
+ *      if blackhole is found
+ * - valid per-nexthop value
+ *      if nexthop of path is valid and connected and a label has been obtained
+ * - invalid value
+ *      if nexthop of path is valid and connected and label has not yet been
+ * obtained
+ */
+static mpls_label_t bgp_mplsvpn_get_path_info_label(struct bgp_path_info *pi,
+						    struct bgp *to_bgp,
+						    struct bgp *from_bgp,
+						    afi_t afi, safi_t safi)
+{
+	struct bgp_nexthop_cache *bnc;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_pfx = NULL;
+	struct prefix nh_resolved = {0};
+	struct bgp *bgp_nexthop = NULL;
+	afi_t nh_afi;
+
+	/* check the next-hop reachability in the
+	 * bgp instance where the next-hop of the entry is
+	 */
+	if (pi->extra && pi->extra->bgp_orig)
+		bgp_nexthop = pi->extra->bgp_orig;
+	else
+		bgp_nexthop = from_bgp;
+
+	nh_afi = BGP_ATTR_NH_AFI(afi, pi->attr);
+
+	if (bgp_find_or_add_nexthop(from_bgp, bgp_nexthop, nh_afi, safi, pi,
+				    NULL, 0, NULL) == 0)
+		/* no valid next-hop, wait for next-hop validity */
+		return MPLS_INVALID_LABEL;
+
+	bnc = pi->nexthop;
+	if (!bnc || bnc->nexthop_num == 0 || bnc->nexthop_num > 1) {
+		/* Invalid next-hop or ECMP routes
+		 * is not compatible with per-nexthop label.
+		 * Fallback to per-vrf label.
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* extract the resolved nexthop given by the BNC nexthop cache
+	 * if nexthop type is IPV4 or IPv6, then use the resolved nexthop
+	 */
+	switch (bnc->nexthop->type) {
+	case NEXTHOP_TYPE_IPV4:
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
+		nh_resolved.family = AF_INET;
+		nh_resolved.prefixlen = IPV4_MAX_BITLEN;
+		IPV4_ADDR_COPY(&nh_resolved.u.prefix4,
+			       &bnc->nexthop->gate.ipv4);
+		nh_pfx = &nh_resolved;
+		break;
+	case NEXTHOP_TYPE_IPV6:
+	case NEXTHOP_TYPE_IPV6_IFINDEX:
+		nh_resolved.family = AF_INET6;
+		nh_resolved.prefixlen = IPV6_MAX_BITLEN;
+		IPV6_ADDR_COPY(&nh_resolved.u.prefix6,
+			       &bnc->nexthop->gate.ipv6);
+		nh_pfx = &nh_resolved;
+		break;
+	case NEXTHOP_TYPE_IFINDEX:
+		/* the nexthop is in bnc->prefix */
```

Comment:
the comment is useless. That is obvious
Say something else

Reply:
done: ```/* the nexthop is direcly connected */```